From 5d91e1b804cd34d8b96275a937380232fb25d451 Mon Sep 17 00:00:00 2001
From: Jeevaka Badrappan <jeevaka.badrappan@intel.com>
Date: Fri, 27 Apr 2012 09:26:10 +0300
Subject: [PORT FROM MAIN]wifi: Reset the InterfaceTable upon failure

BZ: 24583

Since the MOBILE_HIPRI is not a default connection,
ConnectivityService adds the route information of the
established HIPRI connection as a secondary one. Adding
of the route to the SecondaryTable failed. Eventhough the
adding of route information fails, interface name maintained in
the mInterfaceTable structure is not reset. When the call to
enableNat is made by Tethering service, NatController first
checks whether it's a secondary connection. Since the interface
table in the SecondaryTableController is not reset, NatController
tries to do the operations on the invalid secondary interface
resulting in exception thrown and eventually resulting in stopping
the tethering service.

This patch resets the mInterfaceTable upon failure.

Change-Id: I026068405ddf33d0bbcdcdeb693bf86d4ea646dd
Orig-Change-Id: I467238eb1da7d057ba7cc5f16c4a51bf6b22a04d
Category: device enablement
Domain: CWS.WIFI-Common
Origin: upstream
Upstream-Candidate: yes
Signed-off-by: Jeevaka Badrappan <jeevaka.badrappan@intel.com>
---
 SecondaryTableController.cpp |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/SecondaryTableController.cpp b/SecondaryTableController.cpp
index 736b5fe..27ec492 100644
--- a/SecondaryTableController.cpp
+++ b/SecondaryTableController.cpp
@@ -173,6 +173,7 @@ int SecondaryTableController::modifyRoute(SocketClient *cli, const char *action,
         ALOGE("ip route %s failed: %s route %s %s/%d via %s dev %s table %d", action,
                 IP_PATH, action, dest, prefix, gateway, iface, tableIndex+BASE_TABLE_NUMBER);
         errno = ENODEV;
+        mInterfaceTable[tableIndex][0] = 0;
         cli->sendMsg(ResponseCode::OperationFailed, "ip route modification failed", true);
         return -1;
     }
-- 
1.7.4.1

